plugins/xpay/xpay.c

```diff
@@ -1757,14 +1757,99 @@ static const struct plugin_notification notifications[] = {
 	},
 };
 
+/* xpay doesn't have maxfeepercent or exemptfee, so we convert them to
+ * an absolute restriction here.  If we can't, fail and let pay handle
+ * it. */
+static bool calc_maxfee(struct command *cmd,
+			const char **maxfeestr,
+			const char *buf,
+			const jsmntok_t *invstringtok,
+			const jsmntok_t *amount_msattok,
+			const jsmntok_t *exemptfeetok,
+			const jsmntok_t *maxfeepercenttok)
+{
+	u64 maxfeepercent_ppm;
+	struct amount_msat amount, maxfee, exemptfee;
+
+	if (!exemptfeetok && !maxfeepercenttok)
+		return true;
+
+	/* Can't have both */
+	if (*maxfeestr)
+		return false;
+
+	/* If they specify amount easy, otherwise take from invoice */
+	if (amount_msattok) {
+		if (!parse_amount_msat(&amount, buf + amount_msattok->start,
+				       amount_msattok->end - amount_msattok->start))
+			return false;
+	} else {
+		const struct bolt11 *b11;
+		char *fail;
+		const char *invstr;
+
+		/* We need to know total amount to calc fee */
+		if (!invstringtok)
+			return false;
+
+		invstr = json_strdup(tmpctx, buf, invstringtok);
+		b11 = bolt11_decode(tmpctx, invstr, NULL, NULL, NULL, &fail);
+		if (b11 != NULL) {
+			if (b11->msat == NULL)
+				return false;
+			amount = *b11->msat;
+		} else {
+			const struct tlv_invoice *b12;
+			b12 = invoice_decode(tmpctx, invstr, strlen(invstr),
+					     NULL, NULL, &fail);
+			if (b12 == NULL || b12->invoice_amount == NULL)
+				return false;
+			amount = amount_msat(*b12->invoice_amount);
+		}
+	}
+
+	if (maxfeepercenttok) {
+		if (!json_to_millionths(buf,
+					maxfeepercenttok,
+					&maxfeepercent_ppm))
+			return false;
+	} else
+		maxfeepercent_ppm = 500000 / 100;
+
+	if (!amount_msat_fee(&maxfee, amount, 0, maxfeepercent_ppm))
```

Comment:
`maxfeepercent` is a *percent*
```suggestion
	if (!amount_msat_fee(&maxfee, amount, 0, maxfeepercent_ppm/100))
```

Reply:
Good catch!